File: platform_src/ios/Elastos/Classes/AppInfo.swift
 /*
  * Copyright (c) 2018 Elastos Foundation
  *
  * Permission is hereby granted, free of charge, to any person obtaining a copy
  * of this software and associated documentation files (the "Software"), to deal
  * in the Software without restriction, including without limitation the rights
  * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  * copies of the Software, and to permit persons to whom the Software is
  * furnished to do so, subject to the following conditions:
  *
  * The above copyright notice and this permission notice shall be included in all
  * copies or substantial portions of the Software.
  *
  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */
 
import Foundation

@objc(AppInfo)
class AppInfo: NSObject {
    
    @objc static let TID = "tid";
    @objc static let APP_TID = "app_tid";
    @objc static let APP_ID = "app_id";
    @objc static let VERSION = "version";
    @objc static let NAME = "name";
    @objc static let SHORT_NAME = "short_name";
    @objc static let DESCRIPTION = "description";
    @objc static let START_URL = "start_url";
    @objc static let AUTHOR_NAME = "author_name";
    @objc static let AUTHOR_EMAIL = "author_email";
    @objc static let DEFAULT_LOCAL = "default_locale";
    @objc static let BACKGROUND_COLOR = "background_color";
    @objc static let THEME_DISPLAY = "theme_display";
    @objc static let THEME_COLOR = "theme_color";
    @objc static let THEME_FONT_NAME = "theme_font_name";
    @objc static let THEME_FONT_COLOR = "theme_font_color";
    @objc static let INSTALL_TIME = "install_time";
    @objc static let BUILT_IN = "built_in";
    @objc static let REMOTE = "remote";
    @objc static let LAUNCHER = "launcher";
    
    @objc static let SRC = "src";
    @objc static let SIZES = "sizes";
    @objc static let TYPE = "type";
    
    @objc static let PLUGIN = "plugin";
    @objc static let URL = "url";
    @objc static let AUTHORITY = "authority";
    
    @objc dynamic var tid: Int64 = 0;
    @objc dynamic var app_id = "";
    @objc dynamic var version = "";
    @objc dynamic var name = "";
    @objc dynamic var short_name = "";
    @objc dynamic var desc = "";
    @objc dynamic var start_url = "";
    @objc dynamic var author_name = "";
    @objc dynamic var author_email = "";
    @objc dynamic var default_locale = "";
    @objc dynamic var background_color = "";
    @objc dynamic var theme_display = "";
    @objc dynamic var theme_color = "";
    @objc dynamic var theme_font_name = "";
    @objc dynamic var theme_font_color = "";
    @objc dynamic var install_time: Int64 = 0;
    @objc dynamic var built_in = false;
    @objc dynamic var remote = false;
    @objc dynamic var launcher = false;

    @objc static let AUTHORITY_NOEXIST = -1;
    @objc static let AUTHORITY_NOINIT = 0;
    @objc static let AUTHORITY_ASK = 1;
    @objc static let AUTHORITY_ALLOW = 2;
    @objc static let AUTHORITY_DENY = 3;

    @objc var icons = [Icon]();
    @objc var plugins = [PluginAuth]();
    @objc var urls = [UrlAuth]();

    @objc func addIcon(_ src: String, _ sizes: String, _ type: String) {
        icons.append(Icon(src, sizes, type));
    }

    @objc func addPlugin(_ plugin: String, _ authority: Int) {
        plugins.append(PluginAuth(plugin, authority));
    }

    @objc func addUrl(_ url: String, _ authority: Int) {
        urls.append(UrlAuth(url, authority));
    }
 }

 @objc(Icon)
 class Icon: NSObject {
    @objc dynamic var src = "";
    @objc dynamic var sizes = "";
    @objc dynamic var type = "";
    
    init(_ src: String, _ sizes: String, _ type: String) {
        self.src = src;
        self.sizes = sizes;
        self.type = type;
    }
 }

 @objc(PluginAuth)
 class PluginAuth: NSObject {
    @objc dynamic var plugin = "";
    @objc dynamic var authority = AppInfo.AUTHORITY_NOINIT;
    
    init(_ plugin: String, _ authority: Int) {
        self.plugin = plugin;
        self.authority = authority;
    }
 }

 @objc(UrlAuth)
 class UrlAuth: NSObject {
    @objc dynamic var url = "";
    @objc dynamic var authority = AppInfo.AUTHORITY_NOINIT;
    
    init(_ url: String, _ authority: Int) {
        self.url = url;
        self.authority = authority;
    }
 }

File: platform_src/android/app/src/main/java/org/elastos/trinity/runtime/AppInfo.java
 /*
  * Copyright (c) 2018 Elastos Foundation
  *
  * Permission is hereby granted, free of charge, to any person obtaining a copy
  * of this software and associated documentation files (the "Software"), to deal
  * in the Software without restriction, including without limitation the rights
  * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  * copies of the Software, and to permit persons to whom the Software is
  * furnished to do so, subject to the following conditions:
  *
  * The above copyright notice and this permission notice shall be included in all
  * copies or substantial portions of the Software.
  *
  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */


 package org.elastos.trinity.runtime;

 import android.graphics.drawable.Icon;

 import java.util.ArrayList;

 public class AppInfo {

     public static final String TID = "tid";
     public static final String APP_TID = "app_tid";
     public static final String APP_ID = "app_id";
     public static final String VERSION = "version";
     public static final String NAME = "name";
     public static final String SHORT_NAME = "short_name";
     public static final String DESCRIPTION = "description";
     public static final String START_URL = "start_url";
     public static final String AUTHOR_NAME = "author_name";
     public static final String AUTHOR_EMAIL = "author_email";
     public static final String DEFAULT_LOCAL = "default_locale";
     public static final String BACKGROUND_COLOR = "background_color";
     public static final String THEME_DISPLAY = "theme_display";
     public static final String THEME_COLOR = "theme_color";
     public static final String THEME_FONT_NAME = "theme_font_name";
     public static final String THEME_FONT_COLOR = "theme_font_color";
     public static final String INSTALL_TIME = "install_time";
     public static final String BUILT_IN = "built_in";
     public static final String REMOTE = "remote";
     public static final String LAUNCHER = "launcher";

     public static final String SRC = "src";
     public static final String SIZES = "sizes";
     public static final String TYPE = "type";


     public static final String PLUGIN = "plugin";
     public static final String URL = "url";
     public static final String AUTHORITY = "authority";


     public static final int MSG_PARAMS = 0;
     public static final int MSG_RETURN = 1;
     public static final int MSG_URL_AUTHORITY = 2;
     public static final int MSG_PLUGIN_AUTHORITY = 3;

     public long tid;
     public String app_id;
     public String version;
     public String name;
     public String short_name;
     public String description;
     public String start_url;
     public String author_name;
     public String author_email;
     public String default_locale;
     public String background_color;
     public String theme_display;
     public String theme_color;
     public String theme_font_name;
     public String theme_font_color;
     public long   install_time;
     public int    built_in;
     public int    remote;
     public int    launcher;

     public static final int AUTHORITY_NOEXIST = -1;
     public static final int AUTHORITY_NOINIT = 0;
     public static final int AUTHORITY_ASK = 1;
     public static final int AUTHORITY_ALLOW = 2;
     public static final int AUTHORITY_DENY = 3;


     public class Icon {
         public String src = "";
         public String sizes = "";
         public String type = "";

         Icon(String src, String sizes, String type) {
             this.src = src;
             this.sizes = sizes;
             this.type = type;
         }
     }

     public class PluginAuth {
         public String plugin = "";
         public int authority = AUTHORITY_NOINIT;

         PluginAuth(String plugin, int authority) {
             this.plugin = plugin;
             this.authority = authority;
         }
     }

     public class UrlAuth {
         public String url;
         public int authority = AUTHORITY_NOINIT;

         UrlAuth(String url, int authority) {
             this.url = url;
             this.authority = authority;
         }
     }

     public ArrayList<Icon> icons = new ArrayList<Icon>(8);
     public ArrayList<PluginAuth> plugins = new ArrayList<PluginAuth>(20);
     public ArrayList<UrlAuth> urls = new ArrayList<UrlAuth>(20);

     public void addIcon(String src, String sizes, String type) {
         icons.add(new Icon(src, sizes, type));
     }

     public void addPlugin(String plugin, int authority) {
         plugins.add(new PluginAuth(plugin, authority));
     }

     public void addUrl(String url, int authority) {
         urls.add(new UrlAuth(url, authority));
     }
 }
